python/ray/tune/context.py

```diff
@@ -0,0 +1,89 @@
+import threading
+from typing import Any, Dict, Optional
+
+from ray.train._internal import session
+from ray.train.context import TrainContext as TrainV1Context
+from ray.train.context import _copy_doc
+from ray.util.annotations import Deprecated, PublicAPI
+
+# The context singleton on this process.
+_tune_context: Optional["TuneContext"] = None
+_tune_context_lock = threading.Lock()
+
+
+_TRAIN_SPECIFIC_CONTEXT_DEPRECATION_MESSAGE = (
+    "`{}` is deprecated for Ray Tune because there is no concept of worker ranks "
+    "for Ray Tune, so these methods only make sense to use in the context of "
+    "a Ray Train worker."
+)
+
+
+@PublicAPI(stability="stable")
+class TuneContext(TrainV1Context):
+    """Context for Ray Tune training executions."""
+
+    @Deprecated
+    def get_metadata(self) -> Dict[str, Any]:
+        raise DeprecationWarning(
+            "`get_metadata` is deprecated for Ray Tune, as it has never been usable."
```

Comment:
I also need to add `@Deprecated` to the `TrainContext` methods that are getting removed. (`get_trial_*`). However, there is no migration path away from these methods for Ray Train V1.

This highlights a general problem:
* For deprecation warnings in V1, there is sometimes no alternative, since the V1 implementation is staying the same. So there is no clear actionable step for users when they see this message.
* In V2, we can only raise a warning since we don't want to re-implement deprecated behavior.
* So, the deprecation warning just serves as a reminder to try removing these dependencies in user code so that V2 migration later is easier.